File: primes.cpp
#include <cstdio>
#include <cmath>
#include <vector>
#include <algorithm>
#include <cstdlib>
#include <ctime>

using namespace std;

// Some performance optimizations are commented out as an alternative implementation.
// See the comments by Vinicius Miranda for more information:
// http://blog.famzah.net/2010/07/01/cpp-vs-python-vs-perl-vs-php-performance-benchmark/#comment-5347

void get_primes7(int n, vector<int> &res) {
	if (n < 2) return;
	if (n == 2) {
		res.push_back(2);
		return;
	}
	vector<int> s;
	// 1 //s.reserve(n/2); // hint the compiler that we will use "n/2" elements in "s"
	// 2 //s.resize( static_cast<int>(n/2) ); // pre-allocate memory
	// 2 //int j = 0;
	for (int i = 3; i < n + 1; i += 2) {
		s.push_back(i);
		// 2 //s[j] = i;
		// 2 //++j;
	}
	// 2 //s.resize(j);
	int mroot = sqrt(n);
	int half = static_cast<int>(s.size());
	int i = 0;
	int m = 3;
	while (m <= mroot) {
		if (s[i]) {
			int j = static_cast<int>((m*m - 3)*0.5);
			s[j] = 0;
			while (j < half) {
				s[j] = 0;
				j += m;
			}
		}
		i = i + 1;
		m = 2*i + 3;
	}
	// 1 //res.reserve(n/log(n)); // "Prime number theorem" says that we expect this amount of primes
	res.push_back(2);

	/*
	// loop manually
	for (vector<int>::iterator it = s.begin() ; it < s.end(); ++it) {
		if (*it) {
			res.push_back(*it);
		}
	}
	*/

	// use standard methods instead of a loop
	std::vector<int>::iterator pend = std::remove(s.begin(), s.end(), 0);
	res.insert(res.begin() + 1, s.begin(), pend);
}

int main() {
	std::time_t startTime = std::time(NULL);
	std::time_t periodTime = (std::time_t) atoi(std::getenv("RUN_TIME"));

	while ((std::time(NULL) - startTime) < periodTime) {
		vector<int> res;
		get_primes7(10000000, res);
		printf("Found %d prime numbers.\n", (int)res.size());
	}

	return 0;
}

File: primes.rs
use std::env;
use std::time::{Instant, Duration};

const PRIMES_COUNT: u32 = 10000000;

fn get_primes7(count: u32) -> Vec<u32> {
    if count < 2 {
        return vec![];
    } else if count == 2 {
        return vec![2];
    }

    let mut s = Vec::new();
    let mut i = 3;
    while i < count+1 {
        s.push(i);
        i += 2;
    }

    let mroot = (count as f32).sqrt() as u32;
    let half = s.len() as u32;

    let mut i: u32 = 0;
    let mut m: u32 = 3;

    while m <= mroot {
        if s.get(i as usize).is_some() && s[i as usize] != 0 {
            let mut j = (m*m-3)/2;
            s[j as usize] = 0;
            while j < half {
                s[j as usize] = 0;
                j += m;
            }
        }
        i += 1;
        m = 2*i+3;
    }

    let mut res = Vec::new();
    res.push(2);
    res.extend(s.into_iter().filter(|x| *x != 0));
    res
}

fn main() {
    let run_time_secs = match env::var("RUN_TIME") {
        Ok(v) => match v.parse::<u32>() {
            Ok(i) => i,
            Err(err) => panic!("RUN_TIME environment variable error: {}", err),
        },
        Err(err) => panic!("RUN_TIME environment variable error: {}", err),
    };

    let run_time = Duration::new(run_time_secs as u64, 0);
    let start = Instant::now();

    while start.elapsed() < run_time {
	    let primes = get_primes7(PRIMES_COUNT);
        println!("Found {} prime numbers.", primes.len());
    }
}

